# Student Performance Prediction System

## Project Objective

The objective of this project is to build a machine learning model that predicts a student's final academic marks based on academic and personal factors.

### Input Features

- Study Hours
- Attendance
- Previous Score
- Assignment Score
- Gender
- Parent Education

### Target Variable

- Final Marks

### Machine Learning Models

1. Linear Regression
2. Decision Tree Regressor
3. Random Forest Regressor

### Evaluation Metrics

- Mean Absolute Error (MAE)
- Mean Squared Error (MSE)
- R² Score

In [1]:
import pandas as pd
import numpy as np

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)

Pandas version: 2.2.2
NumPy version: 1.26.4


In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
sns.set_theme(style="whitegrid")

plt.rcParams["figure.figsize"] = (8, 5)

In [9]:
df = pd.read_csv(
    "../data/student_performance_dataset.csv"
)

In [10]:
print("Dataset loaded successfully!")

Dataset loaded successfully!


In [11]:
df.head()

,Student_ID,Gender,Study_Hours_per_Week,Attendance_Rate,Past_Exam_Scores,Parental_Education_Level,Internet_Access_at_Home,Extracurricular_Activities,Final_Exam_Score,Pass_Fail
0,S147,Male,31,68.267841,86,High School,Yes,Yes,63,Pass
1,S136,Male,16,78.222927,73,PhD,No,No,50,Fail
2,S209,Female,21,87.525096,74,PhD,Yes,No,55,Fail
3,S458,Female,27,92.076483,99,Bachelors,No,No,65,Pass
4,S078,Female,37,98.655517,63,Masters,No,Yes,70,Pass


In [12]:
df.shape

(708, 10)

In [13]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 708
Number of columns: 10


In [14]:
df.columns.tolist()

['Student_ID',
 'Gender',
 'Study_Hours_per_Week',
 'Attendance_Rate',
 'Past_Exam_Scores',
 'Parental_Education_Level',
 'Internet_Access_at_Home',
 'Extracurricular_Activities',
 'Final_Exam_Score',
 'Pass_Fail']

In [15]:
df.dtypes

Student_ID                     object
Gender                         object
Study_Hours_per_Week            int64
Attendance_Rate               float64
Past_Exam_Scores                int64
Parental_Education_Level       object
Internet_Access_at_Home        object
Extracurricular_Activities     object
Final_Exam_Score                int64
Pass_Fail                      object
dtype: object

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 708 entries, 0 to 707
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Student_ID                  708 non-null    object 
 1   Gender                      708 non-null    object 
 2   Study_Hours_per_Week        708 non-null    int64  
 3   Attendance_Rate             708 non-null    float64
 4   Past_Exam_Scores            708 non-null    int64  
 5   Parental_Education_Level    708 non-null    object 
 6   Internet_Access_at_Home     708 non-null    object 
 7   Extracurricular_Activities  708 non-null    object 
 8   Final_Exam_Score            708 non-null    int64  
 9   Pass_Fail                   708 non-null    object 
dtypes: float64(1), int64(3), object(6)
memory usage: 55.4+ KB


In [17]:
df.describe()

,Study_Hours_per_Week,Attendance_Rate,Past_Exam_Scores,Final_Exam_Score
count,708.000000,708.000000,708.000000,708.000000
mean,26.132768,78.107722,77.871469,58.771186
std,8.877727,13.802802,14.402739,6.705877
min,10.000000,50.116970,50.000000,50.000000
25%,19.000000,67.550094,65.000000,52.000000
50%,27.000000,79.363046,79.000000,59.500000
75%,34.000000,89.504232,91.000000,64.000000
max,39.000000,99.967675,100.000000,77.000000


In [18]:
df.describe(include="object")

,Student_ID,Gender,Parental_Education_Level,Internet_Access_at_Home,Extracurricular_Activities,Pass_Fail
count,708,708,708,708,708,708
unique,500,2,4,2,2,2
top,S339,Female,Bachelors,No,No,Pass
freq,6,375,189,381,361,354


In [19]:
df.isnull().sum()

Student_ID                    0
Gender                        0
Study_Hours_per_Week          0
Attendance_Rate               0
Past_Exam_Scores              0
Parental_Education_Level      0
Internet_Access_at_Home       0
Extracurricular_Activities    0
Final_Exam_Score              0
Pass_Fail                     0
dtype: int64

In [20]:
total_missing = df.isnull().sum().sum()
print("Total missing values:", total_missing)

Total missing values: 0


In [21]:
df.duplicated().sum()

208

In [22]:
duplicate_count = df.duplicated().sum()
duplicate_percentage = (
    duplicate_count / len(df)
) * 100

print("Duplicate rows:", duplicate_count)
print(
    f"Duplicate percentage: "
    f"{duplicate_percentage:.2f}%"
)

Duplicate rows: 208
Duplicate percentage: 29.38%


In [23]:
duplicate_rows = df[
    df.duplicated(keep=False)
]
duplicate_rows.head(10)

,Student_ID,Gender,Study_Hours_per_Week,Attendance_Rate,Past_Exam_Scores,Parental_Education_Level,Internet_Access_at_Home,Extracurricular_Activities,Final_Exam_Score,Pass_Fail
0,S147,Male,31,68.267841,86,High School,Yes,Yes,63,Pass
3,S458,Female,27,92.076483,99,Bachelors,No,No,65,Pass
4,S078,Female,37,98.655517,63,Masters,No,Yes,70,Pass
5,S417,Male,30,84.159193,77,Masters,Yes,No,61,Pass
6,S302,Male,24,89.389494,95,High School,Yes,Yes,61,Pass
8,S044,Female,34,80.863186,94,PhD,No,Yes,65,Pass
16,S327,Male,32,93.031601,78,Bachelors,Yes,Yes,77,Pass
20,S458,Female,27,92.076483,99,Bachelors,No,No,65,Pass
22,S375,Male,36,97.220486,93,Masters,Yes,Yes,65,Pass
24,S207,Female,20,89.075724,92,Bachelors,Yes,Yes,62,Pass


In [24]:
duplicate_rows.shape

(320, 10)

In [25]:
df_cleaned = df.drop_duplicates().copy()

In [26]:
df_cleaned.shape

(500, 10)

In [27]:
df_cleaned.duplicated().sum()

0

In [28]:
categorical_columns = [
    "Gender",
    "Parental_Education_Level",
    "Internet_Access_at_Home",
    "Extracurricular_Activities",
    "Pass_Fail"
]

for column in categorical_columns:
    print(f"\n{column}")
    print(df_cleaned[column].value_counts())


Gender
Gender
Female    256
Male      244
Name: count, dtype: int64

Parental_Education_Level
Parental_Education_Level
High School    135
Bachelors      127
PhD            121
Masters        117
Name: count, dtype: int64

Internet_Access_at_Home
Internet_Access_at_Home
No     263
Yes    237
Name: count, dtype: int64

Extracurricular_Activities
Extracurricular_Activities
No     268
Yes    232
Name: count, dtype: int64

Pass_Fail
Pass_Fail
Fail    354
Pass    146
Name: count, dtype: int64


In [29]:
df_cleaned.to_csv(
    "../data/student_performance_cleaned.csv",
    index=False
)

In [30]:
print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!


## Data Cleaning Summary

The dataset initially contained 708 rows and 10 columns.

### Missing Values

No missing values were found in the dataset. Therefore, no imputation was required during the initial cleaning process.

### Duplicate Records

A total of 208 duplicate rows were identified.

These exact duplicate records were removed to prevent repeated observations from influencing exploratory analysis and machine learning model training.

After duplicate removal:

- Original rows: 708
- Duplicate rows removed: 208
- Final unique rows: 500

### Identifier Column

The `Student_ID` column is an identifier and will not be used as a predictive feature because it does not represent meaningful academic or personal information.

### Target Leakage Prevention

The `Pass_Fail` column will not be used as an input feature when predicting `Final_Exam_Score`, because it may be directly derived from final exam performance and could introduce target leakage.